Support for malformed EMPAD files

The files in https://er-c-data-mgmt.fz-juelich.de/browse/?path=/adhoc/libertem/libertem-test-data/EMPAD/Magical_DataSet#browse don't work with with current EMPAD reader as deployed on moellenstedt.

Loading via `ctx.load("empad", ...)` gives:

```
[...]
File ~/source/LiberTEM/src/libertem/io/dataset/base/file.py:166, in get_array_from_memview()
    162 assert len(mem) % itemsize == 0, \
    163     "len(mem) must fit the dtype"
    164 assert len(mem) // itemsize % self.num_frames == 0, \
    165     "len(mem) must fit the number of frames"
--> 166 assert len(mem) // itemsize // self.num_frames % (
    167     slicing.frame_size + (self._frame_header + self._frame_footer) // itemsize
    168 ) == 0, "len(mem) must fit the sig shape"
    169 arr_uncut = np.frombuffer(mem, dtype=self._native_dtype).reshape(
    170     (self.num_frames, -1)
    171 )
    172 arr = arr_uncut[:, slicing.frame_offset:slicing.frame_offset + slicing.frame_size]

AssertionError: len(mem) must fit the sig shape
```

Probable cause: the raw file has shape `(128, 16)`, and the XML file contains the following:

```
  <scan_parameters mode="search">
    <scan_resolution_x>16</scan_resolution_x>
    <scan_resolution_y>128</scan_resolution_y>
    <scan_center_x>0.5</scan_center_x>
    <scan_size>1.0</scan_size>
    <scan_center_y>0.5</scan_center_y>
  </scan_parameters>
  <scan_parameters mode="acquire">
    <scan_resolution_x>256</scan_resolution_x>
    <scan_resolution_y>256</scan_resolution_y>
    <scan_center_x>0.5</scan_center_x>
    <scan_size>1.0</scan_size>
    <scan_center_y>0.5</scan_center_y>
  </scan_parameters>
```

So the `mode="search"` parameters are actually the correct ones. I'll try to track down how this happens, and we should probably inform TFS about this - but now there are these kind of files around. As a work around for this, we could match the file size with the different scan parameters and select the one that matches.

--- a/docs/source/changelog/bugfix/empad-scan-params.rst
+++ b/docs/source/changelog/bugfix/empad-scan-params.rst
@@ -1,0 +1,7 @@
+[Bugfix] Fix loading of malformed EMPAD data
+============================================
+
+* For some data sets, the actual raw data acquired corresponds to the
+  :code:`mode="search"` scan parameters, and not :code:`mode="acquire"`. This
+  is probably a bug in the acquisition software, but we want to be able to load
+  these files anyways (:issue:`1617`, :pr:`1620`).

--- a/src/libertem/io/dataset/empad.py
+++ b/src/libertem/io/dataset/empad.py
@@ -14,7 +14,18 @@ EMPAD_DETECTOR_SIZE = (128, 128)
 EMPAD_DETECTOR_SIZE_RAW = (130, 128)
 
 
-def get_params_from_xml(path):
+def get_params_from_xml(path, scan_parameters_mode="acquire"):
+    """
+    Parameters
+    ----------
+
+    path
+        The path to the XML file (relative or absolute)
+
+    scan_parameters_mode
+        Either "acquire" or "search" - for some files, the raw
+        data corresponds to the "search" scan parameters.
+    """
     em = ET.parse(path)
     root = em.getroot()
     raw_filename = root.find("raw_file").attrib['filename']
@@ -26,12 +37,12 @@ def get_params_from_xml(path):
 
     typ = root.find("type")
 
+    # assume "scan" if no type is given:
     if typ is None or typ.text == 'scan':
-        # assume "scan":
         scan_parameters = [
             elem
             for elem in root.findall("scan_parameters")
-            if elem.attrib["mode"] == "acquire"
+            if elem.attrib["mode"] == scan_parameters_mode
         ]
 
         node_scan_x = scan_parameters[0].find("scan_resolution_x")
@@ -148,9 +159,12 @@ class EMPADDataSet(DataSet):
         self._path_raw = None
         self._meta = None
 
-    def _init_from_xml(self, path):
+    def _init_from_xml(self, path, shape_fixup: bool = False):
         try:
-            return get_params_from_xml(path)
+            mode = "acquire"
+            if shape_fixup:
+                mode = "search"
+            return get_params_from_xml(path, scan_parameters_mode=mode)
         except Exception as e:
             raise DataSetException(
                 "could not initialize EMPAD file; error: %s" % (
@@ -175,9 +189,32 @@ class EMPADDataSet(DataSet):
             self._filesize = executor.run_function(self._get_filesize)
         except OSError as e:
             raise DataSetException(f"could not open file {self._path_raw}: {str(e)}")
+
+        itemsize = int(np.dtype("float32").itemsize)
+
+        # consistency check: see that the file size matches the detected nav
+        # shape, otherwise, re-run with fixup for #1617
+        if nav_shape_from_XML is not None:
+            expected_size = (
+                prod(EMPAD_DETECTOR_SIZE_RAW) * prod(nav_shape_from_XML) * itemsize
+            )
+            if self._filesize != expected_size:
+                _, nav_shape_from_XML = executor.run_function(
+                    self._init_from_xml, self._path, shape_fixup=True,
+                )
+            # re-run the check...
+            expected_size = (
+                prod(EMPAD_DETECTOR_SIZE_RAW) * prod(nav_shape_from_XML) * itemsize
+            )
+            if self._filesize != expected_size:
+                raise ValueError(
+                    f"RAW data file looks incomplete; filesize={self._filesize} "
+                    f"vs expected size {expected_size}"
+                )
+
         self._image_count = int(
             self._filesize / (
-                int(np.dtype("float32").itemsize) * int(
+                itemsize * int(
                     prod(EMPAD_DETECTOR_SIZE_RAW)
                 )
             )
@@ -186,8 +223,8 @@ class EMPADDataSet(DataSet):
             self._nav_shape = nav_shape_from_XML
         elif self._nav_shape is None and nav_shape_from_XML is None:
             raise ValueError(
-                    "either nav_shape needs to be passed, or path needs to point to the .xml file"
-                )
+                "either nav_shape needs to be passed, or path needs to point to the .xml file"
+            )
         self._nav_shape_product = int(prod(self._nav_shape))
         if nav_shape_from_XML:
             self._image_count = int(prod(nav_shape_from_XML))

--- a/src/libertem/io/dataset/empad.py
+++ b/src/libertem/io/dataset/empad.py
@@ -198,19 +198,25 @@ class EMPADDataSet(DataSet):
             expected_size = (
                 prod(EMPAD_DETECTOR_SIZE_RAW) * prod(nav_shape_from_XML) * itemsize
             )
+            alternate_size = None
+            alternate_nav_shape_from_XML = None
             if self._filesize != expected_size:
-                _, nav_shape_from_XML = executor.run_function(
+                # Extract alternate nav shape with shape_fixup=True
+                _, alternate_nav_shape_from_XML = executor.run_function(
                     self._init_from_xml, self._path, shape_fixup=True,
                 )
-            # re-run the check...
-            expected_size = (
-                prod(EMPAD_DETECTOR_SIZE_RAW) * prod(nav_shape_from_XML) * itemsize
-            )
-            if self._filesize != expected_size:
-                raise ValueError(
-                    f"RAW data file looks incomplete; filesize={self._filesize} "
-                    f"vs expected size {expected_size}"
+                alternate_size = (
+                    prod(EMPAD_DETECTOR_SIZE_RAW) * prod(alternate_nav_shape_from_XML) * itemsize
                 )
+            if self._filesize != expected_size and self._filesize != alternate_size:
+                raise ValueError(
+                    f"RAW data file size mismatch; filesize={self._filesize} "
+                    f"vs expected size {expected_size} for nav {nav_shape_from_XML} "
+                    f"or alternate {alternate_size} for nav {alternate_nav_shape_from_XML}."
+                )
+            elif self._filesize == alternate_size:
+                # apply mitigation
+                nav_shape_from_XML = alternate_nav_shape_from_XML
 
         self._image_count = int(
             self._filesize / (

--- a/tests/io/datasets/test_empad.py
+++ b/tests/io/datasets/test_empad.py
@@ -29,6 +29,9 @@ EMPAD_XML_SERIES = os.path.join(EMPAD_TESTDATA_PATH, 'acquisition_series_pretty.
 EMPAD_XML_BROKEN_PARAMS = os.path.join(
     EMPAD_TESTDATA_PATH, 'Magical_DataSet/acquisition_1/acquisition_1.xml',
 )
+EMPAD_XML_BROKEN_PARAMS_BAD = os.path.join(
+    EMPAD_TESTDATA_PATH, 'Magical_DataSet/acquisition_1/acquisition_1_bad.xml',
+)
 HAVE_EMPAD_TESTDATA = os.path.exists(EMPAD_RAW) and os.path.exists(EMPAD_XML)
 
 pytestmark = pytest.mark.skipif(not HAVE_EMPAD_TESTDATA, reason="need EMPAD testdata")  # NOQA
@@ -83,6 +86,13 @@ def test_new_empad_xml(lt_ctx):
 def test_empad_broken_scan_parameters(lt_ctx):
     ds = lt_ctx.load("empad", path=EMPAD_XML_BROKEN_PARAMS)
     assert ds.shape.to_tuple() == (128, 16, 128, 128)
+
+
+def test_empad_broken_scan_parameters_bad(lt_ctx):
+    with pytest.raises(ValueError) as excinfo:
+        lt_ctx.load("empad", path=EMPAD_XML_BROKEN_PARAMS_BAD)
+    assert "(128, 17)" in str(excinfo.value)
+    assert "(256, 256)" in str(excinfo.value)
 
 
 def test_series_acquisition_xml():

--- a/tests/io/datasets/test_empad.py
+++ b/tests/io/datasets/test_empad.py
@@ -26,6 +26,9 @@ EMPAD_RAW = os.path.join(EMPAD_TESTDATA_PATH, 'scan_11_x4_y4.raw')
 EMPAD_XML = os.path.join(EMPAD_TESTDATA_PATH, 'acquisition_12_pretty.xml')
 EMPAD_XML_2 = os.path.join(EMPAD_TESTDATA_PATH, 'acquisition_12_pretty.xml')
 EMPAD_XML_SERIES = os.path.join(EMPAD_TESTDATA_PATH, 'acquisition_series_pretty.xml')
+EMPAD_XML_BROKEN_PARAMS = os.path.join(
+    EMPAD_TESTDATA_PATH, 'Magical_DataSet/acquisition_1/acquisition_1.xml',
+)
 HAVE_EMPAD_TESTDATA = os.path.exists(EMPAD_RAW) and os.path.exists(EMPAD_XML)
 
 pytestmark = pytest.mark.skipif(not HAVE_EMPAD_TESTDATA, reason="need EMPAD testdata")  # NOQA
@@ -72,12 +75,14 @@ def default_empad_raw():
     return raw_data[:, :, :128, :]
 
 
-def test_new_empad_xml():
-    executor = InlineJobExecutor()
-    ds = EMPADDataSet(
-        path=EMPAD_XML_2,
-    )
-    ds = ds.initialize(executor)
+def test_new_empad_xml(lt_ctx):
+    ds = lt_ctx.load("empad", path=EMPAD_XML_2)
+    assert ds.shape.to_tuple() == (4, 4, 128, 128)
+
+
+def test_empad_broken_scan_parameters(lt_ctx):
+    ds = lt_ctx.load("empad", path=EMPAD_XML_BROKEN_PARAMS)
+    assert ds.shape.to_tuple() == (128, 16, 128, 128)
 
 
 def test_series_acquisition_xml():

